x-pack/elastic-agent/pkg/composable/controller.go

```diff
@@ -8,38 +8,49 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+
 	"reflect"
 	"sort"
 	"sync"
 	"time"
 
 	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/transpiler"
 	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/config"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
 )
 
-// Vars is a context of variables that also contain a list of processors that go with the mapping.
-type Vars struct {
-	Mapping map[string]interface{}
-
-	ProcessorsKey string
-	Processors    []map[string]interface{}
-}
-
 // VarsCallback is callback called when the current vars state changes.
-type VarsCallback func([]Vars)
+type VarsCallback func([]transpiler.Vars)
 
 // Controller manages the state of the providers current context.
-type Controller struct {
+type Controller interface {
+	// Run runs the controller.
+	//
+	// Cancelling the context stops the controller.
+	Run(ctx context.Context, cb VarsCallback) error
+}
+
+// controller manages the state of the providers current context.
+type controller struct {
 	contextProviders map[string]*contextProviderState
 	dynamicProviders map[string]*dynamicProviderState
 }
 
 // New creates a new controller.
-func New(c *config.Config) (*Controller, error) {
-	var providersCfg Config
-	err := c.Unpack(&providersCfg)
+func New(c *config.Config) (Controller, error) {
+	l, err := logger.New("composable")
 	if err != nil {
-		return nil, errors.New(err, "failed to unpack providers config", errors.TypeConfig)
+		return nil, err
+	}
+	l.Info("EXPERIMENTAL - Dynamic Inputs are currently experimental and should not be used in production")
```

Comment:
```suggestion
	l.Info("EXPERIMENTAL - Inputs with variables are currently experimental and should not be used in production")
```

Reply:
When exactly is this logged? What kind of config needs the user to have to see this?

Reply:
This is logged at the start of Elastic Agent when the controller is created. It is always created even without variables in the configuration, so it always logged.